Add a `_freeze` / `_unfreeze` API (#35592)
elastic/elasticsearch: docs/reference/frozen-indices.asciidoc

> [role="xpack"]

Author: @lcawl any pointers how and where we wire this in are appreciated!

Reviewer: In my opinion, this content can all stay in the Elasticsearch Reference.  To that end, I submitted a couple of commits to make it work.  You can preview it here: https://doc-review-63bf6.firebaseapp.com/es2/frozen-indices.html

[role="xpack"]
[testenv="basic"]
[[frozen-indices]]
= Frozen Indices

Elasticsearch indices can require a significant amount of memory available in order to be open and searchable. Yet, not all indices need

> to be writable at the same time and have different access patters over time. For example indices in the time-series or logging use-case

Reviewer 2: ```suggestion
to be writable at the same time and have different access patterns over time. For example indices in the time-series or logging use-case
```

are unlikely to be queried once they age out but still need to be kept around for retention policy purposes.

> In order to keep indices around for a longer time period but at the same time reducing their hardware requirements they can be transitioned

Reviewer 2: ```suggestion
In order to keep indices available and queryable for a longer period but at the same time reduce their hardware requirements they can be transitioned
```

> into a frozen state. Once an index is frozen, all it's transient shard memory aside of mappings and and necessary in-memory structures for

Reviewer 2: ```suggestion
into a frozen state. Once an index is frozen, all of it's transient shard memory (aside from mappings and analyzers)
```

> analysis is moved to persistent storage. This allows for a much higher disk to heap storage ratio on individual nodes. Once an index is

Reviewer 2: ```suggestion
is moved to persistent storage. This allows for a much higher disk to heap storage ratio on individual nodes. Once an index is
```

> frozen their are read-only and require their data-structures required for query execution to be loaded on demand. A search request that hits

Reviewer 2: ```suggestion
frozen, it is made read-only and drops its transient data structures from memory.  These will need to be reloaded on demand (and subsequently dropped) for each search request that targets the frozen index.  A search request that hits
```

> one or more shards from frozen indices will execute it's search through a throttling component that ensures that we never search more than

Reviewer 2: ```suggestion
one or more frozen shards will be executed on a throttled threadpool that ensures that we never search more than
```

> `N` (`1` by default) searches concurrently. This protects nodes from exceeding the available memory due to incoming search requests.

Reviewer 2: ```suggestion
`N` (`1` by default) searches concurrently (see <<search-throttled>>). This protects nodes from exceeding the available memory due to incoming search requests.
```

> In contrast to ordinary open indices, frozen indices are expected to execute slow and are not designed for high query load. Parallelism is

Reviewer 2: ```suggestion
In contrast to ordinary open indices, frozen indices are expected to execute slowly and are not designed for high query load. Parallelism is
```

gained only on a per-node level and loading data-structures on demand is expected to be one or more orders of a magnitude slower than query

> execution on a per shard level. Depending on the index and it's data executing a search on a frozen index is expected to be in the seconds

Reviewer 2: ```suggestion
execution on a per shard level. Depending on the data in an index, a frozen index may execute searches in the seconds to minutes range, when the same index in an unfrozen state may execute the same search request in milliseconds.
```

> or even minutes range compared to milliseconds on a non-frozen index.

Reviewer 2: ```suggestion
```

== Best Practices

> Since frozen indices provide a much higher disk to heap ratio for the expense of latency it is recommended to allocate frozen indices on

Reviewer 2: ```suggestion
Since frozen indices provide a much higher disk to heap ratio at the expense of search latency,  it is advisable to allocate frozen indices to
```

> dedicated nodes to prevent searches on frozen indices influencing traffic on low latency nodes. There is a significant overhead in loading

Reviewer 2: ```suggestion
dedicated nodes to prevent searches on frozen indices influencing traffic on low latency nodes. There is significant overhead in loading
```

> data-structures on demand which can cause page-faults and garbage collections further slowing down query execution.

Reviewer 2: ```suggestion
data-structures on demand which can cause page-faults and garbage collections, which further slow down query execution.
```

> Since indices that are eligible for freezing are likely to not change in the future disk space can be optimized. Many strategies are

Reviewer 2: ```suggestion
Since indices that are eligible for freezing are unlikely to change in the future, disk space can be optimized as described in <<tune-for-disk-usage>>.
```

> outlined in <<tune-for-disk-usage>>.

Reviewer 2: ```suggestion
```

== Freezing and unfreezing an index

> The freeze and unfreeze index APIs allow to freeze an index, and later on

Reviewer 2: ```suggestion
The freeze and unfreeze index APIs are used to freeze and unfreeze an index.
```

> == Freezing and unfreezing an index

Reviewer: In my opinion, this page "Freezing and unfreezing..." page is no longer needed after we move this content to the API reference pages (e.g. https://doc-review-63bf6.firebaseapp.com/es2/freeze-index-api.html and https://doc-review-63bf6.firebaseapp.com/es2/unfreeze-index-api.html)

> unfreeze it. A frozen index has almost no overhead on the cluster (except

Reviewer 2: ```suggestion
A frozen index has almost no overhead on the cluster (except
```

for maintaining its metadata in memory), and is blocked for write operations.

> A frozen index can be unfrozen which will then go through the normal recovery process.

Reviewer 2: ```suggestion
When a frozen index is unfrozen, then index will go through the normal recovery process and become writeable again.
```

The REST endpoint is `/{index}/_freeze` and `/{index}/_unfreeze`. For
example:

[source,js]
--------------------------------------------------
POST /my_index/_freeze

POST /my_index/_unfreeze
--------------------------------------------------
// CONSOLE
// TEST[s/^/PUT my_index\n/]

[IMPORTANT]
================================
 Freezing an index will close the index and reopen it within the same API call. This causes primaries to not be allocated for a short

>  amount of time and causes the cluster to go red until the primaries are allocated again. This limitation might be removed in the future

Reviewer 2: ```suggestion
 amount of time and causes the cluster to go red until the primaries are allocated again. This limitation might be removed in the future
```
```suggestion
 amount of time and causes the cluster to go red until the primaries are allocated again. This limitation might be removed in the future.
```

================================

== Searching a frozen index

Frozen indices are throttled in order limit memory consumptions per node. The number of concurrently loaded frozen indices per node is

> limited by the number of threads in the `search_throttled` <<modules-threadpool,threadpool>> which is `1` by default. At the same time,

Reviewer 2: ```suggestion
limited by the number of threads in the <<search-throttled>> threadpool,  which is `1` by default. 
```

> search requests hitting may indices on a cluster due to concrete lists, expanded wildcards or a date pattern exclude frozen indices by

Reviewer 2: ```suggestion
Search requests will not be executed against frozen indices by default, even if a frozen index is named explicitly. Th is is ```

> default to prevent accidental slowdowns when a frozen index is hit. To include frozen indices a search request must be executed with

Reviewer 2: ```suggestion
to prevent accidental slowdowns by targeting a frozen index by mistake.. To include frozen indices a search request must be executed with
```

> `ignore_throttled=false`.

Reviewer 2: ```suggestion
the query parameter `ignore_throttled=false`.
```

[source,js]
--------------------------------------------------
GET /twitter/_search?q=user:kimchy&ignore_throttled=false
--------------------------------------------------
// CONSOLE
// TEST[setup:twitter]

[IMPORTANT]
================================

> While frozen indices are slow to search, they offer an efficient pre-filter phase. The request parameter `pre_filter_shard_size` specifies

Reviewer 2: ```suggestion
While frozen indices are slow to search, they can be pre-filtered efficiently. The request parameter `pre_filter_shard_size` specifies
```

> a threshold that enforces a round-trip to filter search shards based on query rewriting if the number of shards the search

Reviewer 2: ```suggestion
a threshold that, when exceeded, will enforce a round-trip to pre-filter search shards that cannot possibly match.
```

> request expands to exceeds the threshold. This filter phase can limit the number of shards significantly if for instance a shard can not

Reviewer 2: ```suggestion
This filter phase can limit the number of shards significantly. For instance, if a date range filter is applied, then all indices (frozen or unfrozen) that do not contain documents within the date range can be skipped efficiently.
```

> match any documents based on it's rewrite method ie. if date filters are mandatory to match but the shard bounds and the query are disjoint.

Reviewer 2: ```suggestion
```

> The default value for `pre_filter_shard_size` is `128` while for searching frozen indices it's recommended to set it to `1`. There is no

Reviewer 2: ```suggestion
The default value for `pre_filter_shard_size` is `128` but it's recommended to set it to `1` when searching frozen indices. There is no
```
